File: backend/src/main/java/com/ll/hotel/standard/util/Ut.java
package com.ll.hotel.standard.util;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ll.hotel.global.app.AppConfig;
import com.ll.hotel.global.jwt.dto.JwtProperties;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import javax.crypto.SecretKey;
import java.security.SecureRandom;
import java.util.Date;
import java.util.List;
import java.util.Map;

@Slf4j
public class Ut {

    private final SecretKey secretKey;

    public Ut(JwtProperties jwtProperties) {
        this.secretKey = Keys.hmacShaKeyFor(jwtProperties.getSecret().getBytes());
    }

    public static class str {
        public static boolean isBlank(String str) {
            return str == null || str.trim().isEmpty();
        }
    }

    public static class json {
        private static final ObjectMapper om = AppConfig.getObjectMapper();

        @SneakyThrows
        public static String toString(Object obj) {
            return om.writeValueAsString(obj);
        }

        @SneakyThrows
        public static Map<String, Object> toMap(String jsonStr) {
            return om.readValue(jsonStr, new TypeReference<Map<String, Object>>() {});
        }
    }

    public static class jwt {
        public static String toString(JwtProperties jwtProperties, Map<String, Object> claims) {
            SecretKey secretKey = Keys.hmacShaKeyFor(jwtProperties.getSecret().getBytes());
            Date now = new Date();

            String tokenType = (String) claims.getOrDefault("type", "access");
            long expiration = tokenType.equals("refresh")
                    ? jwtProperties.getRefreshTokenExpiration()
                    : jwtProperties.getAccessTokenExpiration(); // 토큰을 타입으로 구분하여 만료 시간 설정

            return Jwts.builder()
                    .setClaims(claims)
                    .setIssuedAt(now)
                    .setExpiration(new Date(now.getTime() + expiration))
                    .signWith(secretKey)
                    .compact();
        }

        public static Claims getClaims(JwtProperties jwtProperties, String token) {
            SecretKey secretKey = Keys.hmacShaKeyFor(jwtProperties.getSecret().getBytes());
            token = token.replace("Bearer ", "").trim();

            return Jwts.parserBuilder()
                    .setSigningKey(secretKey)
                    .build()
                    .parseClaimsJws(token)
                    .getBody();
        }
    }
  
    public static class random {
        private static final String CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        private static final SecureRandom RANDOM = new SecureRandom();

        public static String generateUID(int length) {
            StringBuilder uid = new StringBuilder(length);

            for (int i = 0; i < length; i++) {
                int randomIndex = RANDOM.nextInt(CHARACTERS.length());
                uid.append(CHARACTERS.charAt(randomIndex));
            }

            return uid.toString();

        }
    }

    public static class list {
        public static boolean hasValue(List<?> list) {
            return list != null && !list.isEmpty();
        }
    }
}

File: backend/src/main/java/com/ll/hotel/standard/util/Ut.kt
package com.ll.hotel.standard.util

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.ll.hotel.global.app.AppConfig
import com.ll.hotel.global.jwt.dto.JwtProperties
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import java.util.Date
import javax.crypto.SecretKey
import kotlin.collections.List

object Ut {
    private lateinit var SecretKey: SecretKey

    fun initSecretKey(jwtProperties : JwtProperties) {
        SecretKey = Keys.hmacShaKeyFor(
            jwtProperties.secret.toByteArray())
    }

    object Str {
        fun isBlank(str: String?) : Boolean = str.isNullOrBlank()
    }

    object Json {
        private val om: ObjectMapper = AppConfig.objectMapper

        fun toString(obj: Any): String = om.writeValueAsString(obj)

        fun toMap(jsonStr: String) :Map<String, Any>{
            return om.readValue(jsonStr, object : TypeReference<Map<String, Any>>() {})
        }
    }

    object Jwt {
        fun toString(jwtProperties: JwtProperties, claims: Map<String, Any>):String {
            val now = Date()
            val secretKey: SecretKey = Keys.hmacShaKeyFor(jwtProperties.secret.toByteArray());

            val tokenType: String = claims["type"] as? String ?: "access"
            val expiration: Long = if (tokenType == "refresh") {
                jwtProperties.refreshTokenExpiration
            } else {
                jwtProperties.accessTokenExpiration
            }

            return Jwts.builder()
                .setClaims(claims)
                .setIssuedAt(now)
                .setExpiration(Date(now.time + expiration))
                .signWith(secretKey)
                .compact()
        }

        fun getClaims(jwtProperties: JwtProperties, token: String): Claims {
            val secretKey: SecretKey = Keys.hmacShaKeyFor(jwtProperties.secret.toByteArray());
            val cleanedToken = token.replace("Bearer ", "").trim()

            return Jwts.parserBuilder()
                .setSigningKey(secretKey)
                .build()
                .parseClaimsJws(cleanedToken)
                .body
        }
    }

    object Random {
        private val characters = ('A'..'Z') + ('a'..'z') + ('0'..'9')

        fun generateUID(length: Int): String {
            return (1..length)
                .map { characters.random() }
                .joinToString("")

        }
    }

    object ListUt {
        fun <T> List<T>?.hasValue(): Boolean {
            return !this.isNullOrEmpty()
        }
    }
}